### ISSUE TITLE
A lot of warnings "member access within misaligned address" when compiling with sanitizers

### ISSUE THREAD

**Issue by qarmin** [nothings/stb#799]
When I compile [Godot Engine](https://github.com/godotengine/godot) which use stb_vorbis.c with sanitizers support:
```
scons p=x11 -j6 use_ubsan=yes use_lsan=yes use_asan=yes
```
and I import ogg files, then this warnings shows
```
thirdparty/misc/stb_vorbis.c:4181:37: runtime error: member access within misaligned address 0x7ffaf9e17064 for type 'struct Mapping', which requires 8 byte alignment
0x7ffaf9e17064: note: pointer points here
  09 09 00 00 01 00 00 00  00 00 00 00 c4 70 e1 f9  fa 7f 00 00 01 00 00 00  00 00 00 00 00 00 00 00
              ^ 
thirdparty/misc/stb_vorbis.c:4182:22: runtime error: member access within misaligned address 0x7ffaf9e2128c for type 'struct stb_vorbis', which requires 8 byte alignment
0x7ffaf9e2128c: note: pointer points here
  fc 01 fc 03 44 ac 00 00  02 00 00 00 d4 51 03 00  00 10 00 00 71 02 00 00  00 00 00 00 00 00 00 00
              ^
thirdparty/misc/stb_vorbis.c:4511:83: runtime error: left shift of 128 by 24 places cannot be represented in type 'int'
thirdparty/misc/stb_vorbis.c:2097:62: runtime error: load of misaligned address 0x7ffaf9e1691c for type 'uint8 *', which requires 8 byte alignment
0x7ffaf9e1691c: note: pointer points here
  fa 7f 00 00 e0 6a e1 f9  fa 7f 00 00 e4 6a e1 f9  fa 7f 00 00 e8 6a e1 f9  fa 7f 00 00 ec 6a e1 f9
              ^ 
```

full log - [wweAAAAAAA.txt](https://github.com/nothings/stb/files/3556475/wweAAAAAAA.txt)

Godot use for now 1.16 version, but this also happens with 1.17

**Comment by nothings** [comment]
Everything defaults to being allocated by malloc. Most likely, either Godot engine overrides the allocator (and isn't staying on 8-byte multiples), or the compiler is being told to pack structures instead of respecting natural alignment. (I mean, maybe there's something in stb_vorbis I'm missing/forgetting, but I don't see anything obvious when looking at it.)

If you want me to look at this, you'll have to reproduce it standalone outside of Godot.

**Comment by qarmin** [comment]
Godot has own memory allocator, but I saw this error only in this module, so I assumed it might be a bug of this repository.
I created an issue https://github.com/godotengine/godot/issues/31776 to get more information

**Comment by pwaller** [comment]
I put my finger on this in the debugger. I littered malloc & godot with asserts of alignment, and everything was aligned until I looked at setup_malloc. I believe the bug is as plain as day in this:

https://github.com/nothings/stb/blob/f67165c2bb2af3060ecae7d20d6f731173485ad0/stb_vorbis.c#L912-L917

`f->alloc.alloc_buffer` is always 8-aligned, but `f->setup_offset` and `sz` are sometimes four-aligned!

I think the `3` in there should be a `7`, which would create the desired alignment, I think. An easy mistake to make because you want to turn 3 bits off ;-). I don't know whether this is valid on different platforms, however.
